### ISSUE TITLE
Missing features in TwUI

### ISSUE THREAD

**Issue by avaidyam** [github/twui#64]
I've listed below what I think are things TwUI should have, can have, and are not that hard to implement. I'm working on a few of these right now, but I've added this list here for the community to add/remove to/from the list, and discuss any changes that could be made. The main twitter/twui fork hasn't been updated in a while now, so I'm posting on the current active fork of TwUI.
## Look and Feel:

I suggest TwUI use appearance containers which allow switchable styles. For example:
- Standard: This appearance would mimic and blend in with Cocoa.
- Light: This would have a brighter scheme, similar to `NSPopover`.
- Dark: This would be a minimal black and white, like `MBProgressHUD`.

The way an appearance container would work, is it would define certain theme colors, fonts, and even drawing and layout blocks. For a control that supports an appearance container, it would have to override its own drawing block/drawRect and layoutSubviews/layout block when you set an appearance container to the object.

A simple example: `TUILabel` could be assigned an appearance container with a primary color black, a secondary color white, an inset for shadow, and a constraint for its position.
A complex example: a `TUIProgressBar` could be assigned a whole drawing block for HUD, that would be a simple rounded rectangle, or a drawing block for a cocoa-style animated progress bar.
## Standard Controls:
- At the moment we have very few controls, compared to Cocoa or UIKit. 
- Add `TUISplitView`.
- Add a full-fledged implementation of a tab bar (vertical or horizontal).
## Navigation Stack:

As far as I know, the reason TwUI doesn't have any navigation-related elements is because of the lack of unification of such elements on Mac; some apps use a breadcrumb bar with back and forward buttons, some use iOS-like navigation headers, and some use a custom derivative. To simplify this, TwUI could allow for a navigation stack engine, and allow the developer to implement it themselves by conforming to a protocol (like `NSTextInputClient`).
- Class `TUINavigationStack` could hold a simple view stack, pushed and popped, with a root view.
- Protocol `TUINavigationPresenter` could have methods such as `-navigationStack:willPush/PopView:` or `-navigationStack:willTransitionFromView:toView:`, to allow for transitions, view updates, and more.

The best part about this is that TwUI doesn't deal with any of the transition animations, views, or data. It just simply allows for pushing and popping a view with "styles" or "transitions". The developer can then write a custom navigation bar, for example, and connect them together. Or perhaps even more complex presentations.

**Comment by CodaFi** [comment]
While I agree with the layout constraints bit (`CALayoutConstraint` wrappers, come on!), I feel a lot of this is misguided, and a little "why not do it yourself"-y.  If you feel there aren't enough controls, write your own and merge them into the master. If you feel `TUITextRenderer` is a bit to heavy for your liking, why not lighten it?    Especially the appledoc bit.  I agree, formal documentation is sparse, but _you_ could change that.

**Comment by avaidyam** [comment]
That's what I was going for - but you see, I can't simply do all this myself. It requires support from the community. So I posted this up here as a "Here's what I think TwUI needs, and here's my plan: change it, do it, or help me?" sort of question. 

And for the record, `CALayoutConstraint` wrappers aren't needed - we have a far better `TUILayoutConstraint` API that allows for value transformation in ways Apple hasn't and perhaps couldn't do. Implemented by yours truly. (See? I'm doing my share too :D)
